### PR TITLE
Early classes for the Distrib Machine Learning Framework.

### PR DESCRIPTION
**Pull request by engineerkatie**
Add initial helper classes used by the ML team.

Update parts of the ML code to use them. Add tests for the comms systems.

### CODE REVIEW

File: apps/muddle-dmlf-chatter/main.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+//------------------------------------------------------------------------------
+
+#include "core/byte_array/const_byte_array.hpp"
+#include "core/byte_array/decoders.hpp"
+#include "core/byte_array/encoders.hpp"
+#include "core/commandline/parameter_parser.hpp"
+#include "core/json/document.hpp"
+#include "core/serializers/main_serializer.hpp"
+#include "variant/variant.hpp"
+//#include "vm/common.hpp"
```
Comment (ejfitzgerald) [inline apps/muddle-dmlf-chatter/main.cpp]: I guess not needed at this point?

File: apps/muddle-dmlf-chatter/main.cpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    auto       t               = TensorType(TensorType::SizeType(2));
+    t.Fill(DataType(sequence_number++));
+    auto r = std::vector<TensorType>();
+    r.push_back(t);
+    interface->pushUpdate(std::make_shared<UpdateTypeForTesting>(r));
+  }
+};
+using Muddle2LearnerNetworker = fetch::dmlf::Muddle2LearnerNetworker;
+
+#pragma clang diagnostic ignored "-Wunused-parameter"
```
Comment (ejfitzgerald) [inline apps/muddle-dmlf-chatter/main.cpp]: Not needed?

File: libs/dmlf/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+include(${FETCH_ROOT_CMAKE_DIR}/BuildTools.cmake)
+
+# Compiler Configuration
+setup_compiler()
+
+# ------------------------------------------------------------------------------
+# Main Library Target
+# ------------------------------------------------------------------------------
+
+# -I/usr/local/Cellar/python/3.7.3/Frameworks/Python.framework/Versions/3.7/include/python3.7m
```
Comment (ejfitzgerald) [inline libs/dmlf/CMakeLists.txt]: Not needed?

File: libs/dmlf/cmake-build-debug/CMakeFiles/3.13.2/CMakeCCompiler.cmake [hunk 4]
```diff
@@ -0,0 +1,69 @@
+set(CMAKE_C_COMPILER "/usr/bin/cc")
```
Comment (ejfitzgerald) [inline libs/dmlf/cmake-build-debug/CMakeFiles/3.13.2/CMakeCCompiler.cmake]: File should not be included in tree

File: libs/dmlf/cmake-build-debug/CMakeFiles/3.13.2/CMakeCXXCompiler.cmake [hunk 5]
```diff
@@ -0,0 +1,80 @@
+set(CMAKE_CXX_COMPILER "/usr/bin/c++")
```
Comment (ejfitzgerald) [inline libs/dmlf/cmake-build-debug/CMakeFiles/3.13.2/CMakeCXXCompiler.cmake]: File should not be included in the tree

File: libs/dmlf/examples/abstract-learner/abstract_learner.cpp [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+using fetch::dmlf::Update;
+
+using namespace std::chrono_literals;
+
+}  // namespace
+
+class FakeLearner : public fetch::dmlf::AbstractLearnerNetworker
+{
+public:
+  void pushUpdate(const std::shared_ptr<UpdateInterface> &update) override
```
Comment (ejfitzgerald) [inline libs/dmlf/examples/abstract-learner/abstract_learner.cpp]: Style 🚓 : I guess we are going to get there with the style over time however, in general methods are `PascalCase`

File: libs/dmlf/examples/abstract-learner/abstract_learner.cpp [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+  void AddPeer(const std::shared_ptr<FakeLearner> &peer)
+  {
+    peers.push_back(peer);
+    peer->peers.push_back(std::shared_ptr<FakeLearner>{this});
+  }
+
+private:
+  using Bytes = AbstractLearnerNetworker::Bytes;
+
+  std::vector<std::shared_ptr<FakeLearner>> peers;
```
Comment (ejfitzgerald) [inline libs/dmlf/examples/abstract-learner/abstract_learner.cpp]: Worth a type alias?

File: libs/dmlf/examples/abstract-learner/abstract_learner.cpp [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+  }
+  std::size_t getPeerCount() const override
+  {
+    return peers.size();
+  }
+
+  void AddPeer(const std::shared_ptr<FakeLearner> &peer)
+  {
+    peers.push_back(peer);
+    peer->peers.push_back(std::shared_ptr<FakeLearner>{this});
```
Comment (ejfitzgerald) [inline libs/dmlf/examples/abstract-learner/abstract_learner.cpp]:
```suggestion
    peer->peers.emplace_back(std::shared_ptr<FakeLearner>{this});
```

File: libs/dmlf/examples/abstract-learner/abstract_learner.cpp [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    {
+      SendMessage(msg, peer);
+    }
+  }
+  std::size_t getPeerCount() const override
+  {
+    return peers.size();
+  }
+
+  void AddPeer(const std::shared_ptr<FakeLearner> &peer)
```
Comment (ejfitzgerald) [inline libs/dmlf/examples/abstract-learner/abstract_learner.cpp]: Think this will cause you issues on the static checks

File: libs/dmlf/include/dmlf/abstract_learner_networker.hpp [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+  }
+
+  AbstractLearnerNetworker(const AbstractLearnerNetworker &other) = delete;
+  AbstractLearnerNetworker &operator=(const AbstractLearnerNetworker &other)  = delete;
+  bool                      operator==(const AbstractLearnerNetworker &other) = delete;
+  bool                      operator<(const AbstractLearnerNetworker &other)  = delete;
+
+protected:
+  std::shared_ptr<ShuffleAlgorithmInterface> alg;                          // used by descendents
+  void                                       NewMessage(const Bytes &msg)  // called by descendents
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/abstract_learner_networker.hpp]: This is where the auto styling causes issue. Worth having lines between definitions

File: libs/dmlf/include/dmlf/abstract_learner_networker.hpp [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+#include "core/byte_array/byte_array.hpp"
+#include "dmlf/queue.hpp"
+#include "dmlf/queue_interface.hpp"
+#include "dmlf/shuffle_algorithm_interface.hpp"
+#include "dmlf/type_map.hpp"
+
+namespace fetch {
+namespace dmlf {
+
+class AbstractLearnerNetworker
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/abstract_learner_networker.hpp:33]: Build times are already an issue. I recommend moving as a rule function definitions to `.cpp` files

File: libs/dmlf/include/dmlf/filepassing_learner_networker.hpp [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+  bool                         operator<(const FilepassingLearnerNetworker &other)  = delete;
+
+protected:
+  static std::string       processNameToTargetDir(const std::string &name);
+  void                     tx(const std::string &target, const Bytes &data);
+  std::vector<std::string> getUpdateNames() const;
+  void                     checkUpdates();
+
+private:
+  std::set<std::string>    processed_updates_;
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/filepassing_learner_networker.hpp]: Type aliases, also is there a benefit from using `std::unordered_set` ?

File: libs/dmlf/include/dmlf/muddle2_learner_networker.hpp [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+  Muddle2LearnerNetworker(const Muddle2LearnerNetworker &other) = delete;
+  Muddle2LearnerNetworker &operator=(const Muddle2LearnerNetworker &other)  = delete;
+  bool                     operator==(const Muddle2LearnerNetworker &other) = delete;
+  bool                     operator<(const Muddle2LearnerNetworker &other)  = delete;
+
+protected:
+  CertificatePtr CreateIdentity();
+  CertificatePtr LoadIdentity(const std::string &privkey);
+
+  CertificatePtr ident;
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/muddle2_learner_networker.hpp]: Attributes are usually suffixed with `_`

File: libs/dmlf/include/dmlf/queue.hpp [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+
+#include "dmlf/queue_interface.hpp"
+#include "dmlf/update.hpp"
+#include "dmlf/update_interface.hpp"
+
+namespace fetch {
+namespace dmlf {
+
+template <typename T>
+class Queue : public QueueInterface
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/queue.hpp:33]: You might find the "core/containters/queue.hpp" might be useful here. It does not have priority support current but it is unclear thow important that is here in this case

File: libs/dmlf/include/dmlf/queue.hpp [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+  using QueueUpdates = std::priority_queue<UpdatePtr, std::vector<UpdatePtr>, std::greater<>>;
+
+  Queue()           = default;
+  ~Queue() override = default;
+
+  void PushNewMessage(Bytes msg) override
+  {
+    auto update = std::make_shared<UpdateType>();
+    update->deserialise(msg);
+    std::cout << "[Queue] Got update: " << update->TimeStamp() << std::endl;
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/queue.hpp]: Logging macro?

File: libs/dmlf/include/dmlf/queue.hpp [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+
+  std::size_t size() const override
+  {
+    Lock l{updates_m_};
+    return updates_.size();
+  }
+
+  std::shared_ptr<UpdateType> getUpdate()
+  {
+    Lock l{updates_m_};
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/queue.hpp]:
```suggestion
    FETCH_LOCK(updates_m_);
```

File: libs/dmlf/include/dmlf/queue.hpp [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+    std::cout << "[Queue] Got update: " << update->TimeStamp() << std::endl;
+    {
+      Lock l{updates_m_};
+      updates_.push(update);
+    }
+  }
+
+  std::size_t size() const override
+  {
+    Lock l{updates_m_};
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/queue.hpp]:
```suggestion
    FETCH_LOCK(updates_m_);
```

File: libs/dmlf/include/dmlf/type_map.hpp [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+    if (iter != map.end())
+    {
+      throw std::runtime_error{"Type already registered with name '" + iter->second + "'"};
+    }
+    // map[getTypeId<KeyType>()] = std::forward<ValueType>(value);
+    map[getTypeId<KeyType>()] = value;
+  }
+
+  template <typename KeyType>
+  inline static int getTypeId()
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/type_map.hpp]: This looks like a round about way to basically run `typid` to get a `std::type_info`. Maybe using that would provide a cleaner interface?

File: libs/dmlf/include/dmlf/update.hpp [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+  Update(const Update &other) = delete;
+  Update &operator=(const Update &other)  = delete;
+  bool    operator==(const Update &other) = delete;
+  bool    operator<(const Update &other)  = delete;
+
+protected:
+private:
+  static TimeStampType CurrentTime()
+  {
+    return static_cast<TimeStampType>(std::chrono::duration_cast<std::chrono::milliseconds>(
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/update.hpp:101]: You might want to consider looking to the `fetch-moment` library for clocks here. @n-hutton will be doing something similar in the future. The advantage of the `fetch-moment` library is that the clocks are adjustable from tests

File: libs/dmlf/include/dmlf/update_interface.hpp [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+           std::to_string(TimeStamp());
+  }
+
+  // Queue ordering
+  bool operator>(const UpdateInterface &other) const
+  {
+    return TimeStamp() > other.TimeStamp();
+  }
+
+  virtual ~UpdateInterface() = default;
```
Comment (ejfitzgerald) [inline libs/dmlf/include/dmlf/update_interface.hpp]: Keep constructors / destructors in the same place?

File: libs/dmlf/src/muddle2_learner_networker.cpp [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+
+  server->Add(1, proto.get());
+
+  auto config_peers = doc.root()["peers"];
+
+  for (std::size_t peer_number = 0; peer_number < config_peers.size(); peer_number++)
+  {
+    if (peer_number != instance_number)
+    {
+      peers.push_back(config_peers[peer_number]["pub"].As<std::string>());
```
Comment (ejfitzgerald) [inline libs/dmlf/src/muddle2_learner_networker.cpp]:
```suggestion
      peers.emplace_back(config_peers[peer_number]["pub"].As<std::string>());
```

File: libs/dmlf/src/muddle2_learner_networker.cpp [hunk 22]
```diff
@@ -0,0 +1,186 @@
...
+void Muddle2LearnerNetworker::pushUpdateType(const std::string &                     type,
+                                             const std::shared_ptr<UpdateInterface> &update)
+{
+  auto client = std::make_shared<RpcClient>("Client", mud->GetEndpoint(), 1, 1);
+  auto data   = update->serialise(type);
+
+  PromiseList promises;
+  promises.reserve(20);
+
+#pragma clang diagnostic ignored "-Wunused-variable"
```
Comment (ejfitzgerald) [inline libs/dmlf/src/muddle2_learner_networker.cpp]: Not needed?

File: libs/dmlf/src/shuffle_algorithm_interface.cpp [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "dmlf/shuffle_algorithm_interface.hpp"
```
Comment (ejfitzgerald) [inline libs/dmlf/src/shuffle_algorithm_interface.cpp:19]: Is this file needed?

File: libs/dmlf/src/update.cpp [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+//
+//------------------------------------------------------------------------------
+
+#include "dmlf/update.hpp"
+
+namespace fetch {
+namespace dmlf {
+
+}  // namespace dmlf
+}  // namespace fetch
```
Comment (ejfitzgerald) [inline libs/dmlf/src/update.cpp]: Populate file with implementation?

File: libs/dmlf/src/vm_wrapper_etch.cpp [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+  vm_->AttachOutputDevice(VM::STDOUT, outputStream_);
+  status_ = VmWrapperInterface::COMPILED;
+  return errors;
+}
+void VmWrapperEtch::Execute(const std::string &entrypoint, const Params & /*params*/)
+{
+  status_ = VmWrapperInterface::RUNNING;
+  std::string        error;
+  fetch::vm::Variant output;
+  outputStream_ = std::stringstream();  // Clear the output stream
```
Comment (ejfitzgerald) [inline libs/dmlf/src/vm_wrapper_etch.cpp:58]:
```suggestion
  outputStream_ = std::ostringstream{};  // Clear the output stream
```

**Review by ejfitzgerald — CHANGES_REQUESTED** [review]
Some initial comments

File: libs/dmlf/include/dmlf/queue.hpp (hunk 14 above)
Comment (baykaner) [reply libs/dmlf/include/dmlf/queue.hpp:33]: Priority is actually fairly helpful in this context. That's because these queues will presumably hold training 'Updates'. As a general rule, older Updates will be less useful; thus a controller might choose to ignore them or apply them based on their position in the queue.

File: libs/dmlf/tests/muddle/muddle_test.cpp [hunk 26]
```diff
@@ -0,0 +1,215 @@
...
+
+using LNBase  = fetch::dmlf::AbstractLearnerNetworker;
+using LN      = fetch::dmlf::Muddle2LearnerNetworker;
+using LNBaseP = std::shared_ptr<LNBase>;
+using LNP     = std::shared_ptr<LN>;
+
+using UpdateTypeForTesting = fetch::dmlf::Update<TensorType>;
+using UpdatePayloadType    = UpdateTypeForTesting::PayloadType;
+
+// static constexpr char const *NETWORK_A_PUBLIC_KEY =
```
Comment (baykaner) [inline libs/dmlf/tests/muddle/muddle_test.cpp]: commented out keys?

File: scripts/fetchai_code_quality/internal/static_analysis.py [hunk 27]
```diff
@@ -53,7 +53,7 @@ def read_static_analysis_yaml(project_root, build_root, yaml_file_path):
             file_path = d['FilePath']
             absolute_path = file_path if isabs(
                 file_path) else abspath(join(build_root, file_path))
-            assert isfile(absolute_path)
+            #assert isfile(absolute_path), "Not a file: {}".format(absolute_path)
```
Comment (baykaner) [inline scripts/fetchai_code_quality/internal/static_analysis.py]: presumably this should not be commented out in master.
Either this assert should be here or it shouldn't?

Comment (wkaluza) [reply scripts/fetchai_code_quality/internal/static_analysis.py]: This assert is valid and should be restored.

**Review by wkaluza — CHANGES_REQUESTED** [review]

File: apps/muddle-dmlf-chatter/main.cpp (hunk 2 above)
Comment (engineerkatie) [reply apps/muddle-dmlf-chatter/main.cpp]: deleted.

File: libs/dmlf/CMakeLists.txt (hunk 3 above)
Comment (engineerkatie) [reply libs/dmlf/CMakeLists.txt]: deleted

File: apps/muddle-dmlf-chatter/main.cpp (hunk 1 above)
Comment (engineerkatie) [reply apps/muddle-dmlf-chatter/main.cpp]: rmvd

File: libs/dmlf/cmake-build-debug/CMakeFiles/3.13.2/CMakeCXXCompiler.cmake (hunk 5 above)
Comment (engineerkatie) [reply libs/dmlf/cmake-build-debug/CMakeFiles/3.13.2/CMakeCXXCompiler.cmake]: rmvd

File: libs/dmlf/cmake-build-debug/CMakeFiles/3.13.2/CMakeCCompiler.cmake (hunk 4 above)
Comment (engineerkatie) [reply libs/dmlf/cmake-build-debug/CMakeFiles/3.13.2/CMakeCCompiler.cmake]: rmvd

File: libs/dmlf/examples/abstract-learner/abstract_learner.cpp (hunk 6 above)
Comment (engineerkatie) [reply libs/dmlf/examples/abstract-learner/abstract_learner.cpp]: Doing

File: libs/dmlf/include/dmlf/abstract_learner_networker.hpp (hunk 11 above)
Comment (engineerkatie) [reply libs/dmlf/include/dmlf/abstract_learner_networker.hpp:33]: doing

File: libs/dmlf/include/dmlf/muddle2_learner_networker.hpp (hunk 13 above)
Comment (engineerkatie) [reply libs/dmlf/include/dmlf/muddle2_learner_networker.hpp]: doing.

File: libs/dmlf/include/dmlf/queue.hpp (hunk 16 above)
Comment (engineerkatie) [reply libs/dmlf/include/dmlf/queue.hpp]: Think i have all these now.

File: libs/dmlf/include/dmlf/update_interface.hpp (hunk 20 above)
Comment (engineerkatie) [reply libs/dmlf/include/dmlf/update_interface.hpp]: done

File: libs/dmlf/src/muddle2_learner_networker.cpp (hunk 21 above)
Comment (engineerkatie) [reply libs/dmlf/src/muddle2_learner_networker.cpp]: done

File: libs/dmlf/include/dmlf/queue.hpp (hunk 17 above)
Comment (engineerkatie) [reply libs/dmlf/include/dmlf/queue.hpp]: as above

File: libs/dmlf/include/dmlf/queue.hpp (hunk 15 above)
Comment (engineerkatie) [reply libs/dmlf/include/dmlf/queue.hpp]: deleted

File: libs/dmlf/src/update.cpp (hunk 24 above)
Comment (engineerkatie) [reply libs/dmlf/src/update.cpp]: done

File: libs/dmlf/tests/muddle/muddle_test.cpp (hunk 26 above)
Comment (engineerkatie) [reply libs/dmlf/tests/muddle/muddle_test.cpp]: my bad, gone

File: scripts/fetchai_code_quality/internal/static_analysis.py (hunk 27 above)
Comment (engineerkatie) [reply scripts/fetchai_code_quality/internal/static_analysis.py]: Oops.
Soz.

File: libs/dmlf/src/muddle2_learner_networker.cpp (hunk 22 above)
Comment (engineerkatie) [reply libs/dmlf/src/muddle2_learner_networker.cpp]: Not even sure what it's doing there...
done

File: libs/dmlf/src/shuffle_algorithm_interface.cpp (hunk 23 above)
Comment (engineerkatie) [reply libs/dmlf/src/shuffle_algorithm_interface.cpp:19]: move impls into it.

File: libs/dmlf/include/dmlf/type_map.hpp (hunk 18 above)
Comment (lrahmani) [reply libs/dmlf/include/dmlf/type_map.hpp]: @ejfitzgerald
that is true.
changing the TypeMap class to use `std::type_index` 
check the Notes section  https://en.cppreference.com/w/cpp/language/typeid

File: libs/dmlf/include/dmlf/update.hpp (hunk 19 above)
Comment (lrahmani) [reply libs/dmlf/include/dmlf/update.hpp:101]: @ejfitzgerald which clock do you recommend : `GetClock()` or `CreateAdjustableClock()`?

Comment (lrahmani) [reply libs/dmlf/include/dmlf/update.hpp:101]: Understood. I'll move code to use `GetClock()`

File: libs/dmlf/examples/abstract-learner/abstract_learner.cpp (hunk 7 above)
Comment (engineerkatie) [reply libs/dmlf/examples/abstract-learner/abstract_learner.cpp]: done

File: libs/dmlf/examples/abstract-learner/abstract_learner.cpp (hunk 8 above)
Comment (engineerkatie) [reply libs/dmlf/examples/abstract-learner/abstract_learner.cpp]: done

File: libs/dmlf/src/muddle2_learner_networker.cpp (hunk 21 above)
Comment (engineerkatie) [reply libs/dmlf/src/muddle2_learner_networker.cpp]: done

File: libs/dmlf/src/muddle2_learner_networker.cpp (hunk 22 above)
Comment (engineerkatie) [reply libs/dmlf/src/muddle2_learner_networker.cpp]: edleted

File: libs/dmlf/src/vm_wrapper_etch.cpp (hunk 25 above)
Comment (engineerkatie) [reply libs/dmlf/src/vm_wrapper_etch.cpp:58]: ok

Comment (engineerkatie) [reply libs/dmlf/src/vm_wrapper_etch.cpp:58]: Actually, this doesn't work because we later want getline() to run on it.

File: libs/dmlf/include/dmlf/filepassing_learner_networker.hpp (hunk 12 above)
Comment (engineerkatie) [reply libs/dmlf/include/dmlf/filepassing_learner_networker.hpp]: done

File: libs/dmlf/examples/abstract-learner/abstract_learner.cpp (hunk 9 above)
Comment (engineerkatie) [reply libs/dmlf/examples/abstract-learner/abstract_learner.cpp]: Altered

**Comment by engineerkatie** [comment]
Superceded
